internals/secrethub/run.go

```diff
@@ -112,66 +116,156 @@ func (cmd *RunCommand) Register(r command.Registerer) {
+
+	// Pass all signals to child process
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals)
+
+	go func() {
+		select {
+		case s := <-signals:
+			err := command.Process.Signal(s)
+			if err != nil && !strings.Contains(err.Error(), "process already finished") {
+				fmt.Fprintln(os.Stderr, ErrSignalFailed(err))
+			}
+		case <-done:
+			signal.Stop(signals)
+			return
+		}
+	}()
 
+	commandErr := command.Wait()
+	done <- true
+
+	if !cmd.noMasking {
+		err = maskedStdout.Flush()
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		err = maskedStderr.Flush()
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+	}
+
+	if commandErr != nil {
+		// Check if the program exited with an error
+		exitErr, ok := commandErr.(*exec.ExitError)
+		if ok {
+			waitStatus, ok := exitErr.Sys().(syscall.WaitStatus)
+			if ok {
+				// Return the status code returned by the process
+				os.Exit(waitStatus.ExitStatus())
+				return nil
+			}
+
+		}
+		return commandErr
+	}
+
+	return nil
+}
+
+// sourceEnvironment returns the environment of the subcommand, with all the secrets sourced
+// and the secret values that need to be masked.
+func (cmd *RunCommand) sourceEnvironment() ([]string, []string, error) {
 	osEnv, passthroughEnv := parseKeyValueStringsToMap(cmd.osEnv)
 
+	envSources := []EnvSource{}
+
 	referenceEnv := newReferenceEnv(osEnv)
 	envSources = append(envSources, referenceEnv)
 
 	// TODO: Validate the flags when parsing by implementing the Flag interface for EnvFlags.
 	flagSource, err := NewEnvFlags(cmd.envar)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 	envSources = append(envSources, flagSource)
 
 	if cmd.envFile == "" {
 		const defaultEnvFile = "secrethub.env"
-		_, err := os.Stat(defaultEnvFile)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				return fmt.Errorf("could not read default run env-file %s: %s", defaultEnvFile, err)
-			}
-		} else {
+		_, err := cmd.osStat(defaultEnvFile)
+		if err == nil {
 			cmd.envFile = defaultEnvFile
+		} else if !os.IsNotExist(err) {
+			return nil, nil, fmt.Errorf("could not read default run env-file %s: %s", defaultEnvFile, err)
 		}
 	}
 
 	if cmd.envFile != "" {
 		templateVariableReader, err := newVariableReader(osEnv, cmd.templateVars)
 		if err != nil {
-			return err
+			return nil, nil, err
 		}
 
 		if !cmd.dontPromptMissingTemplateVar {
 			templateVariableReader = newPromptMissingVariableReader(templateVariableReader, cmd.io)
 		}
 
-		raw, err := ioutil.ReadFile(cmd.envFile)
+		raw, err := cmd.readFile(cmd.envFile)
 		if err != nil {
-			return ErrCannotReadFile(cmd.envFile, err)
+			return nil, nil, ErrCannotReadFile(cmd.envFile, err)
 		}
 
 		parser, err := getTemplateParser(raw, cmd.templateVersion)
 		if err != nil {
-			return err
+			return nil, nil, err
 		}
 
-		envFile, err := ReadEnvFile(cmd.envFile, templateVariableReader, parser)
+		envFile, err := ReadEnvFile(cmd.envFile, bytes.NewReader(raw), templateVariableReader, parser)
```

Comment:
Modified `ReadEnvFile` to take the file name and a reader to its contents, instead of opening the env file again.